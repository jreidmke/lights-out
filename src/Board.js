import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows, ncols, chanceLightStartsOn }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    // TODO: create array-of-arrays of true/false values

    for(let i = 0; i < nrows; i++) {
      let colArr = [];
      for(let i = 0; i < ncols; i++) {
        colArr.push(Math.random() > chanceLightStartsOn ? true : false);
      }
      initialBoard.push(colArr);
    }
    console.log(initialBoard);
    return initialBoard;
  }

  function hasWon() {
    // TODO: check the board in state to determine whether the player has won.
    return (board.flat().filter(x => x == false)).length === 0;
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      // TODO: Make a (deep) copy of the oldBoard
      let newBoard = [...board];
      // TODO: in the copy, flip this cell and the cells around it
      flipCell(y, x, newBoard);
      flipCell(y, x - 1, boardCopy);
      flipCell(y - 1, x, boardCopy);
      flipCell(y + 1, x, boardCopy);
      flipCell(y, x + 1, boardCopy);
      // TODO: return the copy
      return newBoard;
    });
  }

  // if the game is won, just show a winning msg & render nothing else

  // TODO

  // make table board

  // TODO

  return(
    <Cell/>
  )
}

Board.defaultProps = {
  nrows: 5,
  ncols: 5,
  chanceLightStartsOn: .2
}

export default Board;
