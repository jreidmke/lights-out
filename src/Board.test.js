import React from "react";
import { fireEvent, render } from "@testing-library/react";
import Board from "./Board";

// smoke test
it("renders without crashing", function() {
  render(<Board />);
});

it("matches snapshot", function() {
    const { asFragment } = render(<Board chanceLightStartsOn={1}/>)
    expect(asFragment()).toMatchSnapshot();
})

it("displays winning text when game over", function() {
    const { getByText } = render(<Board chanceLightStartsOn={1}/>);
    const h1 = getByText('Congratulations. You have won the game!');
    expect(h1).toBeInTheDocument();
})