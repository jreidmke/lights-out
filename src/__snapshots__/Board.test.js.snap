// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Congratulations. You have won the game!
    </h1>
  </div>
</DocumentFragment>
`;
